AX: align AccessibilityRole::Switch with ::Checkbox

annevk opened the pull request:
#### 5ce55c698227f6ac484867f0b741cb9b8c73f4c7
<pre>
AX: align AccessibilityRole::Switch with ::Checkbox
<a href="https://bugs.webkit.org/show_bug.cgi?id=263008">https://bugs.webkit.org/show_bug.cgi?id=263008</a>
rdar://116805287

Reviewed by Tyler Wilcock.

A switch should essentially be a checkbox except that it doesn&apos;t
support the intermediate/mixed state.

As such, this makes the following changes:

- We make switch support the required attribute.
- We make interaction regions support switches.

* LayoutTests/accessibility/aria-required-expected.txt:
* LayoutTests/accessibility/aria-required.html:
* LayoutTests/inspector/dom/getAccessibilityPropertiesForNode-expected.txt:
* LayoutTests/inspector/dom/getAccessibilityPropertiesForNode.html:
* LayoutTests/interaction-region/aria-roles-expected.txt:
* LayoutTests/interaction-region/aria-roles.html:
* LayoutTests/platform/gtk/inspector/dom/getAccessibilityPropertiesForNode-expected.txt:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::supportsRequiredAttribute const):
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::shouldAllowAccessibilityRoleAsPointerCursorReplacement):

Canonical link: <a href="https://commits.webkit.org/269311@main">https://commits.webkit.org/269311@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/179d3741eb299467a90512401e04ca73d7573133

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22190 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22397 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23262 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24086 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20546 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22438 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26652 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22718 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21567 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22423 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22138 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19254 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24937 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19197 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20110 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26364 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20206 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20335 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24224 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20775 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17680 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20126 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/19919 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5284 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24337 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20723 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cookiecrook reviewed (commented):
Will you also confirm `computedrole` with a new subtest in WPT?
https://github.com/web-platform-tests/wpt/blob/master/html-aam/roles.html

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -269,7 +269,7 @@ LayoutRect AccessibilityNodeObject::checkboxOrRadioRect() const
 {
-    if (isCheckboxOrRadio())
+    if (isCheckboxOrRadio() || isSwitch())
         return checkboxOrRadioRect();
```
The purpose `checkboxOrRadioRect()` is to unify the rect of the control with the rect of its associated label. But right now this only works with native `<label for="id-of-control">Foo label</label>`, or containing labels:

```html
<label>
   Foo bar label for switch
   <div role="switch"></div>
</label>
```

But I'm wondering if either of these methods of pairing labels are "allowed" or recommended patterns for a `role="switch"`? Certainly for `<input type=checkbox switch>`, but `role="switch"` I'm less confident in.

@cookiecrook what do you think?

annevk replied:
Hmm, but the logic here also does this for `<label><span role=checkbox></span></label>`, right? Perhaps it shouldn't be using `isCheckboxOrRadio()` to begin with?

twilco replied:
Yes you're right, it would behave that way for role=checkbox, but arguably shouldn't. Let's remove this part and we'll address it separately later.

annevk replied:
#19048

cookiecrook replied:
Yes label for/id should only work on native inputs, not ARIA controls.

annevk committed:
AX: align AccessibilityRole::Switch with ::Checkbox
https://bugs.webkit.org/show_bug.cgi?id=263008
rdar://116805287

Reviewed by Tyler Wilcock.

A switch should essentially be a checkbox except that it doesn't
support the intermediate/mixed state.

As such, this makes the following changes:

- We make switch support the required attribute.
- We make interaction regions support switches.

* LayoutTests/accessibility/aria-required-expected.txt:
* LayoutTests/accessibility/aria-required.html:
* LayoutTests/inspector/dom/getAccessibilityPropertiesForNode-expected.txt:
* LayoutTests/inspector/dom/getAccessibilityPropertiesForNode.html:
* LayoutTests/interaction-region/aria-roles-expected.txt:
* LayoutTests/interaction-region/aria-roles.html:
* LayoutTests/platform/gtk/inspector/dom/getAccessibilityPropertiesForNode-expected.txt:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::supportsRequiredAttribute const):
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::shouldAllowAccessibilityRoleAsPointerCursorReplacement):

Canonical link: https://commits.webkit.org/269311@main
